### ISSUE TITLE
Implement end-to-end tests

### ISSUE THREAD

**Issue by LukSteib** [sidestream-tech/unified-auctions-ui#282]
### Goal

Implement end-to-end test

### Context
Implement end-to-end test based on the extensive investigation and discussion here: #277 

### Tasks

- [ ] Implement end-to-end test

**Comment by KirillDogadin-std** [comment]
__State of things so far__

Well, here we have our first fun part (non-straightforward problem) of testing our app with cypress

As you could see on pic [1], we fire tons of requests in order to populate the table. The cypress apparently is not loading these fast enough during its own execution and does not put info in the table in reasonable time. The loading time is at least 20 seconds.

The cypress log looks like pic [2]

And after waiting for long time (20+ seconds) i can see pic [3] that is expected on the test network.


__assets__

Pic [1]

![Image](https://user-images.githubusercontent.com/59374892/171352314-ae032834-b442-49c6-bbf5-29fa6920edda.png)

Pic [2]

![Image](https://user-images.githubusercontent.com/59374892/171352915-cf5b5edf-2334-47cb-b9cb-6e9e45a8f1c2.png)


Pic [3]

![Image](https://user-images.githubusercontent.com/59374892/171353007-bcdca72d-0778-4abd-b6cc-f18c3a5157f3.png)

**Comment by valiafetisov** [comment]
The problem is not specific to the cypress. Since hardhat is forking blockchain on a specific block, if needs "historical data", and is much slower to fetch (or maybe it's hardhat itself which makes it slow). The problem can be resolved:

- By providing a [timeout](https://docs.ethers.io/v5/api/utils/web/#ConnectionInfo) to the provider (just add `timeout` key to the options, besides `url`) here https://github.com/sidestream-tech/unified-auctions-ui/blob/8d976c1c74618c3179059c311d232d2dcf2ee949/core/src/provider.ts#L8
- By keeping hardhat cache in a preserved volume inside CI, so we don't refetch those historical blocks on every run (hardhat is nicely caching it locally out-of-the-box already, we just need to keep that cache between runs)

**Comment by KirillDogadin-std** [comment]
> By providing a [timeout](https://docs.ethers.io/v5/api/utils/web/#ConnectionInfo) to the provider

this will result into the empty auction table. that is never populated and then the e2e flow is not possible to complete - even if there is an auction there will not be a button in the table to click since we did not fetch the entry due to timeout..

___ 

so to be clear: in browser the table appears instantly, then i spin up cypress and it still takes ages to load (or timeouts and does not ever display a thing if i set it to). Therefore, if the problem/chokepoint is hardhat, i still fail to understand why in the browser and in the cypress-tool the execution speed is so much different.

**Comment by KirillDogadin-std** [comment]
> The screenshots from above (Pic [1]) 

sorry for confusion, but the picture there was only to demonstrate the number of requests to the service that contains the auction info (localhost:port is not informative). 

> Have you tried switching to the localhost network in the locally running ui?

all my above claims hold because i was testing this on this local setup all along

> Otherwise I am confused to why cypress requests would be slower or "never timeout"? 

i also dont really understand how and why cypress is that slow, that is one of the reasons why the overall discussion is happening :)

but the requests do timeout when i set them to do so. When i dont  - they dont (pretty much like `requests` in python).

>  I still assume that's what is happening

final test;
 - open cypress and run the test that gets to the page that throws these requests,
   - see them loading
 - go to tab with `http://localhost:3000/collateral?network=localhost` - the app, reload, see the table.
 - go to cypress
   - it still loads
   - it shows the table after some time.

**Comment by valiafetisov** [comment]
Since cypress is [based on chromium](https://github.com/cypress-io/cypress/blob/566a7b1feb0fc1a8f1ccf83a23d8ad7a94409a6b/package.json#L145), it should act same as your browser, unless there is some meaningless configuration defaults. Maybe you can increase any of their built-in timeouts? https://docs.cypress.io/guides/references/configuration#Timeouts

Btw, what happens with the requests, does any of them resolve? Is it actually "slow" or just never resolved? Maybe there is just no access to localhost whatsoever, and you need to whitelist it?

> When i dont - they dont (pretty much like requests in python).

They do, but default ethers.js timeout is long, 120 seconds if I remember correctly.

**Comment by KirillDogadin-std** [comment]
> Is it actually "slow" or just never resolved? Maybe there is just no access to localhost whatsoever, and you need to whitelist it?

cypress eventually displays the table after 40-120 seconds. so they actually resolve.

> They do, but default ethers.js timeout is long, 120 seconds if I remember correctly.

ok , then they resolve within this time. but if `isLoading` parameter of the table depends on __all__ requests being resolved , then we are as slow as the slowest+latest request.

**Comment by KirillDogadin-std** [comment]
https://github.com/cypress-io/cypress/issues/8156

this also raises the same concern.

___

> Maybe you can increase any of their built-in timeouts?

i don't see how increased timeouts would help since it's always the requests that execute forever.

**Comment by KirillDogadin-std** [comment]
I'd like to kick off the discussion about the ci.

Here's the proposal to the implementation of it, pls discuss how this complies with the current drone setup (which is quite sophisticated):

 as in https://www.cypress.io/blog/2019/05/02/run-cypress-with-a-single-docker-command/ we use the docker image that is already setup to run the tests.

Along with it the following images have to run:
  - frontend
  - hardhat

[In the drone](https://github.com/sidestream-tech/unified-auctions-ui/blob/8d976c1c74618c3179059c311d232d2dcf2ee949/.drone.yml#L222) then this will require the corresponding configuration:
 - running services (frontend, hardhat)
 - set up environment variables
 - something that was not thought of yet ...

**Comment by valiafetisov** [comment]
Yep, sounds good.
- We can use newly built frontend "production" image to start frontend as a service (so not a special "test build")
- Use cyprus test image to run a test (but I'm not sure if we can share a folder with code or if we have to build a new frontend "test image" based on the cyprus)

> something that was not thought of yet

Yes, the other unresolved question is how to cache hardhat cache inside drone. The endpoint hardhat provides initially fetches required data from the API that is a) slow b) limits amount of requests per day. But hardhat by default caches those requests and never re-requests them as soon as the cache is available. The cache is a folder and it's path [can be configured](https://hardhat.org/config#path-configuration). We probably need to add some permanent storage for it.

**Comment by KirillDogadin-std** [comment]
So to be clear: while saying "permanent storage" you're implying that the drone stores this, right?

**Comment by valiafetisov** [comment]
Yes, probably there is a way to mount a volume to preserve things between CI runs

**Comment by KirillDogadin-std** [comment]
Preservation according to the discussion with @deusavalon caching solution is not straightforward and 

after taking a look at https://plugins.drone.io/plugins/volume-cache, apparently has to be done via a plugin.

If one wants to avoid using the plugin, one option could be to [make the repo trusted](https://laszlo.cloud/the-ultimate-droneci-caching-guide). The link here also contains an instruction on how to use the plugin.

**Comment by valiafetisov** [comment]
Ok, let's start without cache then and see 1) how much time it takes without cache 2) how many requests it fires (and if we need to buy some quota then)
